src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,170 @@
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list != null) {
+            for (T element : list) {
+                add(element);
+            }
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (findByIndex(index)).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = findByIndex(index);
+        T oldValue = node.item;
+        node.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = findByIndex(index);
+        T result = node.item;
+        unlink(node);
+        size--;
+        return result;
     }
 
     @Override
     public boolean remove(T object) {
+        if (findByItem(object) != null) {
+            unlink(findByItem(object));
+            size--;
+            return true;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Cann`t find index " + index);
+        }
+    }
+
+    private void checkToAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Cann`t find index " + index);
+        }
+    }
+
+    private Node<T> findByIndex(int index) {
+        checkIndex(index);
+        Node<T> currentNode;
+        if (size / 2 > index) {
+            currentNode = first;
+            for (int i = 1; i <= index; i++) {
+                currentNode = currentNode.next;
+            }
+        } else {
+            currentNode = last;
+            for (int i = size - 1; i > index; i--) {
+                currentNode = currentNode.prev;
+            }
+        }
+        return currentNode;
+    }
+
+    private Node<T> findByItem(T value) {
+        Node<T> current = first;
+        while (current != null) {
+            if (current.item == value || (current.item != null && (current.item).equals(value))) {
+                return current;
+            }
+            current = current.next;
+        }
+        return null;
+    }
+
+    private void unlink(Node<T> node) {
+        if (node.prev != null && node.next != null) {
+            node.next.prev = node.prev;
+            node.prev.next = node.next;
+        }
+        if (node.next == null) {
+            last = node.prev;
+            if (node.prev != null) {
+                node.prev.next = null;
+            }
+        }
+        if (node.prev == null) {
+            first = node.next;
+            if (node.next != null) {
+                node.next.prev = null;
+            }
```

Comment:
this method looks a little too complicated, let's try to use two if-else blocks here
tip: we can create temporary nodes node.prev and node.next, then do two checks:
if (prev == null) {...}
else {...}
if (next == null) {...}
else {...}